File: Data Structures & Algorithms/validate-parentheses/submission-12.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        values = {}
        open_p = '({['

        values[')'] = '('
        values['}'] = '{'
        values[']'] = '['

        for value in s:
            if value in open_p:
                print("here!!!")
                stack.append(value)
            
            elif len(stack) == 0:
                return False
            elif len(stack) != 0 and values.get(value) != stack[-1]:
                return False
            else:
                stack.pop()
        
        return len(stack) == 0

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-12.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if(root == null){
            return false;
        }


        if(isValid(root, subRoot)){
            System.out.println("Here is the result");
            return true;
        }

        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }

    public boolean isValid(TreeNode root, TreeNode subRoot){
        if(root == null && subRoot == null){
            return true;
        }

        if(root != null && subRoot != null && root.val == subRoot.val){
            return isValid(root.left, subRoot.left) && isValid(root.right, subRoot.right);
        }

        return false;
    }
}
